README.md
# Barito Flow

![Build Status](https://travis-ci.org/BaritoLog/barito-flow.svg?branch=master)

Component for handling flow of logs within a cluster. Support 2 modes:
- **Producer**, for receiving logs and forwarding it to Kafka
- **Consumer**, for consuming from kafka and forwarding it to Elasticsearch

Compatible with both gRPC and REST API. The use of REST API is optional and is implemented by using 
[gRPC-gateway](https://github.com/grpc-ecosystem/grpc-gateway) which work like a reverse-proxy server 
to translate RESTful HTTP API into gRPC. Barito flow infrastructure consists of producer and consumer.

Barito flow producer will turn on gRPC server and optionally REST gateway reverse proxy server.
It will automatically create Kafka topic for the log if not exist yet.
gRPC messages and services are declared in [barito-proto](https://github.com/vwidjaya/barito-proto) repository.

Barito flow consumer will firstly create a topic event and generates the workers. Then based on the logs send, 
each topic inside this event topic will be created a cluster consumer separately. This cluster consumer will 
store the logs to Elasticsearch by calling a single store or bulk store Elasticsearch API. If the process has 
failed, Elasticsearch will halt all the workers and retry again after some backoff period. The halted workers 
only continue when the failed process success on the retry attempt.

## Development Setup 

Fetch and build the project.
```sh
git clone https://github.com/BaritoLog/barito-flow
cd barito-flow
go build
```

Generate mock classes.
```sh
mockgen -source=flow/leaky_bucket.go -destination=mock/leaky_bucket.go -package=mock
mockgen -source=flow/kafka_admin.go -destination=mock/kafka_admin.go -package=mock
mockgen -source=flow/Vendor/github.com/sarama/sync_producer.go -destination=mock/sync_producer.go -package=mock
```

### Running Test Stack using Docker Compose

First, you need to install Docker on your local machine. Then you can run `docker-compose`:

```sh
$ docker-compose -f docker/docker-compose.yml up -d
```

This will pull Elasticsearch, Kafka, and build producer and consumer image. The ports
are mapped as if they are running on local machine.

## Producer Mode

Responsible for:
- Receive logs by exposing an HTTP endpoint
- Produce message to kafka cluster

After the project is built, run:
```sh
./barito-flow producer

# or
./barito-flow p
```

Endpoints using REST gateway:

POST /produce
```json
  {
    "context": {
        "kafka_topic": "kafka_topic",
        "kafka_partition": 1,
        "kafka_replication_factor": 1,
        "es_index_prefix": "es_index_prefix",
        "es_document_type": "es_document_type",
        "app_max_tps": 100,
        "app_secret": "app_secret"
    },
    "timestamp": "optional timestamp here",
    "content": {
        "hello": "world",
        "key": "value",
        "num": 100
    },
  }
```

POST /produce_batch
```json
  {
    "context": {
        "kafka_topic": "kafka_topic",
        "kafka_partition": 1,
        "kafka_replication_factor": 1,
        "es_index_prefix": "es_index_prefix",
        "es_document_type": "es_document_type",
        "app_max_tps": 100,
        "app_secret": "app_secret"
    },
    "items": [
        {
            "content": {
                "timber_num": 1
            }
        },
        {
            "content": {
                "timber_num": 2
            }
        }
    ]
  }
```

These environment variables can be modified to customize its behaviour.

| Name| Description | ENV | Default Value  |
| ---|---|---|---|
| ConsulUrl | Consul URL | BARITO_CONSUL_URL | |
| ConsulKafkaName | Kafka service name in consul | BARITO_CONSUL_KAFKA_NAME | kafka |
| KafkaBrokers | Kafka broker addresses (CSV). Get from env is not available in consul | BARITO_KAFKA_BROKERS | localhost:9092 |
| KafkaMaxRetry | Number of retry to connect to kafka during startup | BARITO_KAFKA_MAX_RETRY | 0 (unlimited) |
| KafkaRetryInterval | Interval between retry connecting to kafka (in seconds) | BARITO_KAFKA_RETRY_INTERVAL | 10 |
| ServeRestApi | Toggle for REST gateway api | BARITO_PRODUCER_REST_API | true |
| ProducerAddressGrpc | gRPC Server Address | BARITO_PRODUCER_GRPC| :8082 |
| ProducerAddressRest | REST Server Address | BARITO_PRODUCER_REST| :8080 |
| ProducerMaxRetry | Set kafka setting max retry | BARITO_PRODUCER_MAX_RETRY | 10 |
| ProducerMaxTps | Producer rate limit trx per second | BARITO_PRODUCER_MAX_TPS | 100 |
| ProducerRateLimitResetInterval | Producer rate limit reset interval (in seconds) | BARITO_PRODUCER_RATE_LIMIT_RESET_INTERVAL | 10 |

## Consumer Mode

Responsible for:
- Consume logs from kafka
- Commit logs to elasticsearch

After the project is built, run:
```sh
./barito-flow Consumer

# or
./barito-flow c
```

These environment variables can be modified to customize its behaviour.

| Name| Description | ENV | Default Value  |
| ---|---|----|----|
| ConsulUrl | Consul URL | BARITO_CONSUL_URL | |
| ConsulKafkaName  | Kafka service name in consul | BARITO_CONSUL_KAFKA_NAME | kafka |
| ConsulElasticsearchName | Elasticsearch service name in consul | BARITO_CONSUL_ELASTICSEARCH_NAME | elasticsearch |
| KafkaBrokers | Kafka broker addresses (CSV). Get from env is not available in consul | BARITO_KAFKA_BROKERS| "127.0.0.1:9092,192.168.10.11:9092" |
| KafkaGroupID | kafka consumer group id | BARITO_KAFKA_GROUP_ID | barito-group |
| KafkaMaxRetry | Number of retry to connect to kafka during startup | BARITO_KAFKA_MAX_RETRY | 0 (unlimited) |
| KafkaRetryInterval | Interval between retry connecting to kafka (in seconds) | BARITO_KAFKA_RETRY_INTERVAL | 10 |
| ElasticsearchUrls | Elastisearch addresses. Get from env is not available in consul | BARITO_ELASTICSEARCH_URLS | "http://127.0.0.1:9200,http://192.168.10.11:9200" |
| EsIndexMethod | BulkProcessor / SingleInsert | BARITO_ELASTICSEARCH_INDEX_METHOD | BulkProcessor |
| EsBulkSize | BulkProcessor bulk size | BARITO_ELASTICSEARCH_BULK_SIZE | 100 |
| EsFlushIntervalMs | BulkProcessor flush interval (ms) | BARITO_ELASTICSEARCH_FLUSH_INTERVAL_MS | 500 |
| PrintTPS | print estimated consumed every second | BARITO_PRINT_TPS | false |
| PushMetricUrl | push metric api url | BARITO_PUSH_METRIC_URL|   |
| PushMetricInterval | push metric interval | BARITO_PUSH_METRIC_INTERVAL | 30s |

**NOTE**  
These following variables will be ignored if `BARITO_ELASTICSEARCH_INDEX_METHOD` is set to `SingleInsert`

- `BARITO_ELASTICSEARCH_BULK_SIZE`
- `BARITO_ELASTICSEARCH_FLUSH_INTERVAL_MS`

### Changelog

See [CHANGELOG.md](CHANGELOG.md)

### Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md)

### License

MIT License, See [LICENSE](LICENSE).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.13.5]

### Changed

- Fix error cardinality while push metric to Prometheus

## [0.13.4]

### Changed

- Add metrics when barito-flow failed connect to ES or Kafka
- Adjust log message when barito-flow failed connect to ES or Kafka

## [0.13.3]

### Changed

- Make barito-flow-consumer compatible with both ES 6.x and ES 7.x
- Add basic auth for xpack-security feature
- Remove index settings when creating new index
- Set document type to `_doc`

## [0.13.2]

### Changed

- Add more metrics on producer
- Fixed: return early on error, before calling any method on nil value

## [0.13.1]

### Changed

- Reorder config to fetch kafka and elasticsearch value

## [0.13.0]

### Changed

- Support GRPC with gateway
- Add instrumentation for prometheus

## [0.12.2]

### Changed

- Enable consul to get multi elasticsearch urls, if urls not available then using environment variable

## [0.12.1]

### Fixed

- Update instru to the latest version, to avoid race condition

## [0.12.0]

### Added

- Add the ability to use bulk indexing when committing logs to Elasticsearch

### Changed

- Only create Elasticsearch client once, instead of recreating it before every commit

## [0.11.8]

### Changed

- Retry connecting to kafka during startup if kafka cluster is not available (configurable)

## [0.11.7]

### Changed

- Enable toggle-able verbose mode

## [0.11.6]

### Changed

- Set round robin strategy as default for consumer rebalancing, also allow it to be configured

## [0.11.5]

### Changed

- Recreate cluster admin whenever barito-flow needs to create a new topic

## [0.11.4]

### Changed

- Update rate limiter behaviour for produce batch, now it counts rate per line in batch instead of per batch request

## [0.11.3]

### Changed

- Lower metadata refresh frequency

## [0.11.2]

### Fixed

- Refresh metadata before creating new topic (see [here](https://github.com/Shopify/sarama/issues/1162))

## [0.11.1]

### Fixed

- Rate limiter on produce batch should count per request, not per log

## [0.11]

### Added

- Ability to process batch logs

## [0.10.1]

### Fixed

- Use updated instru that avoid race condition problem on counter

## [0.10.0]

### Added

- Implement Elasticsearch backoff functionality

## [0.9.0]

### Changed

- Upgrade sarama to 1.19.0
- Use sarama ClusterAdmin for creating topic

## [0.8.5]

### Changed

- Produce logs if flow has to return 5xx errors

## [0.8.4]

### Changed

- Use updated instru that avoid race condition problem

## [0.8.3]

### Changed

- Also send application secret from context when sending metrics

## [0.8.2]

### Changed

- Send metrics to market

## [0.8.1]

### Fixed

- Fix bug when checking for new events

## [0.8.0]

### Added

- Support for multiple topics and indexes

## [0.7.0]

### Added

- Graceful Shutdown

## [0.6.0]

### Changed

- Using consul to get kafka brokers, if consul not available then using environment variable
- Using consul to get elasticsearch url, if url not available then using environment variable

## [0.5.0]

### Changed

- Rate limit trx per second (by default 100)

## [0.4.0]

### Changed

- Rename receiver to producer
- Rename forwarder to consumer


cmds/env.go
package cmds

import (
	"github.com/BaritoLog/go-boilerplate/envkit"
)

// TODO: rename env.go to config.go
// TODO: rename function name using config prefix
// TODO: using log.info() get config value and config source

const (
	EnvKafkaBrokers        = "BARITO_KAFKA_BROKERS"
	EnvKafkaGroupID        = "BARITO_KAFKA_GROUP_ID"
	EnvKafkaConsumerTopics = "BARITO_KAFKA_CONSUMER_TOPICS"
	EnvKafkaProducerTopic  = "BARITO_KAFKA_PRODUCER_TOPIC"
	EnvElasticsearchUrl    = "BARITO_ELASTICSEARCH_URL"

	EnvPushMetricUrl      = "BARITO_PUSH_METRIC_URL"
	EnvPushMetricToken    = "BARITO_PUSH_METRIC_TOKEN"
	EnvPushMetricInterval = "BARITO_PUSH_METRIC_INTERVAL"

	EnvProducerAddress  = "BARITO_PRODUCER_ADDRESS"
	EnvProducerMaxRetry = "BARITO_PRODUCER_MAX_RETRY"
	EnvProducerMaxTPS   = "BARITO_PRODUCER_MAX_TPS"

	EnvConsulUrl               = "BARITO_CONSUL_URL"
	EnvConsulKafkaName         = "BARITO_CONSUL_KAFKA_NAME"
	EnvConsulElasticsearchName = "BARITO_CONSUL_ELASTICSEARCH_NAME"
)

var (
	DefaultConsulKafkaName   = "kafka"
	DefaultElasticsearchName = "elasticsearch"
	DefaultKafkaBrokers      = []string{"localhost:9092"}

	DefaultElasticsearchUrl    = "http://localhost:9200"
	DefaultKafkaConsumerTopics = []string{"consumer-topic"}
	DefaultKafkaGroupID        = "barito-group"
	DefaultPushMetricUrl       = "http://localhost:3000/api/increase_log_count"
	DefaultPushMetricToken     = ""
	DefaultPushMetricInterval  = "30s"
	DefaultProducerAddress     = ":8080"
	DefaultProducerMaxRetry    = 10
	DefaultKafkaProducerTopic  = "producer-topic"
	DefaultProducerMaxTPS      = 100
)

func getKafkaBrokers() (brokers []string) {
	brokers, err := consulKafkaBroker()
	if err != nil {
		brokers = envkit.GetSlice(EnvKafkaBrokers, ",", DefaultKafkaBrokers)
	}
	return
}

func getElasticsearchUrl() (url string) {
	url, err := consulElasticsearchUrl()
	if err != nil {
		url = envkit.GetString(EnvElasticsearchUrl, DefaultElasticsearchUrl)
	}
	return
}

func getConsulElasticsearchName() string {
	return envkit.GetString(EnvConsulElasticsearchName, DefaultElasticsearchName)
}

func getKafkaConsumerTopics() []string {
	return envkit.GetSlice(EnvKafkaConsumerTopics, ",", DefaultKafkaConsumerTopics)
}

func getKafkaGroupId() string {
	return envkit.GetString(EnvKafkaGroupID, DefaultKafkaGroupID)
}

func getPushMetricUrl() string {
	return envkit.GetString(EnvPushMetricUrl, DefaultPushMetricUrl)
}

func getPushMetricToken() string {
	return envkit.GetString(EnvPushMetricToken, DefaultPushMetricToken)
}

func getPushMetricInterval() string {
	return envkit.GetString(EnvPushMetricInterval, DefaultPushMetricInterval)
}

func getProducerAddress() string {
	return envkit.GetString(EnvProducerAddress, DefaultProducerAddress)
}

func getProducerMaxRetry() int {
	return envkit.GetInt(EnvProducerMaxRetry, DefaultProducerMaxRetry)
}

func getKafkaProducerTopic() string {
	return envkit.GetString(EnvKafkaProducerTopic, DefaultKafkaProducerTopic)
}

func getProducerMaxTPS() int {
	return envkit.GetInt(EnvProducerMaxTPS, DefaultProducerMaxTPS)
}
